File: src/main/java/org/processmining/models/time_driven_behavior/NoiseToken.java
package org.processmining.models.time_driven_behavior;

import org.deckfour.xes.model.XTrace;
import org.processmining.models.MovementResult;
import org.processmining.models.organizational_extension.Resource;
import org.processmining.utils.TimeDrivenLoggingSingleton;

/**
 * @author Ivan Shugurov
 *         Created  30.07.2014
 */
public class NoiseToken extends TimeDrivenToken
{
    private final Object modelActivity;


    public NoiseToken(Object modelActivity, Resource resource, long timestamp)
    {
        super(null, resource, timestamp);
        this.modelActivity = modelActivity;
    }

    @Override
    public TimeDrivenToken copyTokenWithNewTimestamp(long newTimestamp)
    {
        return new NoiseToken(modelActivity, getResource(), newTimestamp);
    }

    @Override
    public MovementResult move(XTrace trace)
    {
        MovementResult<TimeDrivenToken> movementResult = new MovementResult<TimeDrivenToken>();
        movementResult.addConsumedExtraToken(this);
        registerEvent(trace);
        return movementResult;
    }

    private void registerEvent(XTrace trace)
    {
        if (getResource() == null)
        {
            TimeDrivenLoggingSingleton.timeDrivenInstance().log(trace, modelActivity, getTimestamp(), true);
        }
        else
        {
            TimeDrivenLoggingSingleton.timeDrivenInstance().logCompleteEventWithResource(trace, modelActivity, getResource(), getTimestamp());
        }
    }

}

File: console/src/main/kotlin/org/processmining/models/time_driven_behavior/NoiseToken.kt
package org.processmining.models.time_driven_behavior

import org.deckfour.xes.model.XTrace
import org.processmining.models.MovementResult
import org.processmining.models.organizational_extension.Resource
import org.processmining.utils.TimeDrivenLoggingSingleton

/**
 * @author Ivan Shugurov
 * Created  30.07.2014
 */
class NoiseToken(
        private val modelActivity: Any,
        timestamp: Long,
        resource: Resource?
) : TimeDrivenToken(null, timestamp, resource) {
    
    override fun copyTokenWithNewTimestamp(newTimestamp: Long): TimeDrivenToken {
        return NoiseToken(modelActivity, newTimestamp, resource)
    }
    
    override fun move(trace: XTrace): MovementResult<*> {
        val movementResult = MovementResult<TimeDrivenToken>()
        movementResult.addConsumedExtraToken(this)
        registerEvent(trace)
        return movementResult
    }
    
    private fun registerEvent(trace: XTrace) {
        
        val logger = TimeDrivenLoggingSingleton.timeDrivenInstance
        
        if (resource == null) {
            logger.log(trace, modelActivity, timestamp, true)
        } else {
            logger.logCompleteEventWithResource(trace, modelActivity, resource, timestamp)
        }
    }
    
}
